AX: Implement TextUnit::Sentence (current) off the main thread

hoffmanjoshua opened the pull request:
#### 57e3f9bf4e22d7ff0ba3f08344782a277fbeca38
<pre>
AX: Implement TextUnit::Sentence (current) off the main thread
<a href="https://bugs.webkit.org/show_bug.cgi?id=281182">https://bugs.webkit.org/show_bug.cgi?id=281182</a>
<a href="https://rdar.apple.com/134579696">rdar://134579696</a>

Reviewed by Tyler Wilcock.

This patch supports calculating the current TextUnit::Sentence on the accessibility thread. Building
off of the work done for supporting LeftWord/RightWord, I reuse that existing infrastructure, and
abstract out the calculations for word and sentence text units into lambdas.

The partially ported test, text-marker-sentence-nav.html, only verifies the current-sentence API.
This test will be expanded once more Sentence APIs are implemented on the accessibility thread
(see &quot;FIXME&quot;s in the test).

* LayoutTests/accessibility/ax-thread-text-apis/text-marker-sentence-nav-expected.txt: Added.
* LayoutTests/accessibility/ax-thread-text-apis/text-marker-sentence-nav.html: Added.
* Source/WebCore/accessibility/AXTextMarker.cpp:
(WebCore::previousSentenceStartFromPosition):
(WebCore::nextSentenceEndFromPosition):
(WebCore::AXTextMarker::findMarker const):
(WebCore::AXTextMarker::sentenceRange const):
* Source/WebCore/accessibility/AXTextMarker.h:
(WebCore::AXTextMarker::previousSentenceStart const):
(WebCore::AXTextMarker::nextSentenceEnd const):
* Source/WebCore/accessibility/mac/WebAccessibilityObjectWrapperMac.mm:
(-[WebAccessibilityObjectWrapper textMarkerRangeAtTextMarker:forUnit:]):

Canonical link: <a href="https://commits.webkit.org/284951@main">https://commits.webkit.org/284951@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/2145234063d6daf72411e0b36b1f59959b304abb

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/70980 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/50392 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/23753 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/75084 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/22185 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/73096 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/58191 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/22003 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/56137 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/14614 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/74046 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/45785 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/61185 "Passed tests") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/36584 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [⏳ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/iOS-17-Simulator-WPT-WK2-Tests-EWS "Waiting to run tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/18619 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/20526 "Built successfully") | 
| | [⏳ 🧪 api-ios](https://ews-build.webkit.org/#/builders/API-Tests-iOS-Simulator-EWS "Waiting to run tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/18979 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/76830 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/15212 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/62/builds/18178 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/63886 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/15256 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/61244 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/63847 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/15716 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/11928 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/119/builds/5573 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/46193 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/968 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/47265 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/48548 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/47007 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

twilco reviewed (approved):
🙌

twilco commented on Source/WebCore/accessibility/AXTextMarker.cpp:
```diff
@@ -853,6 +867,29 @@ AXTextMarker AXTextMarker::findMarker(AXDirection direction, AXTextUnit textUnit
+                if (direction == AXDirection::Previous && end <= objectBorder && end != -1)
+                    resultMarker = AXTextMarker(currentObject->treeID(), currentObject->objectID(), end);
+                else if (direction == AXDirection::Next && end != -1 && end >= objectBorder)
+                    resultMarker = AXTextMarker(currentObject->treeID(), currentObject->objectID(), end - objectBorder);
```
A good follow-up change to this PR would be making a new `AXTextMarker` constructor that takes an `AXCoreObject` and an offset so we aren't repeating `AXTextMarker(currentObject->treeID(), currentObject->objectID(), offset)` all over the place (instead all these callsites could look like `AXTextMarker(currentObject, offset)`)

hoffmanjoshua replied:
I was noticing that as well... I can definitely follow up with that!

hoffmanjoshua committed:
AX: Implement TextUnit::Sentence (current) off the main thread
https://bugs.webkit.org/show_bug.cgi?id=281182
rdar://134579696

Reviewed by Tyler Wilcock.

This patch supports calculating the current TextUnit::Sentence on the accessibility thread. Building
off of the work done for supporting LeftWord/RightWord, I reuse that existing infrastructure, and
abstract out the calculations for word and sentence text units into lambdas.

The partially ported test, text-marker-sentence-nav.html, only verifies the current-sentence API.
This test will be expanded once more Sentence APIs are implemented on the accessibility thread
(see "FIXME"s in the test).

* LayoutTests/accessibility/ax-thread-text-apis/text-marker-sentence-nav-expected.txt: Added.
* LayoutTests/accessibility/ax-thread-text-apis/text-marker-sentence-nav.html: Added.
* Source/WebCore/accessibility/AXTextMarker.cpp:
(WebCore::previousSentenceStartFromPosition):
(WebCore::nextSentenceEndFromPosition):
(WebCore::AXTextMarker::findMarker const):
(WebCore::AXTextMarker::sentenceRange const):
* Source/WebCore/accessibility/AXTextMarker.h:
(WebCore::AXTextMarker::previousSentenceStart const):
(WebCore::AXTextMarker::nextSentenceEnd const):
* Source/WebCore/accessibility/mac/WebAccessibilityObjectWrapperMac.mm:
(-[WebAccessibilityObjectWrapper textMarkerRangeAtTextMarker:forUnit:]):

Canonical link: https://commits.webkit.org/284951@main
